File: cli/internal/commands/clean.go
package commands

import (
	"context"
	"fmt"
	"os"

	"github.com/41vi4p/expo-builder-local/cli/internal/color"
	"github.com/41vi4p/expo-builder-local/cli/internal/config"
	"github.com/41vi4p/expo-builder-local/cli/internal/dockerapi"
)

const cleanUsage = `ebl clean [options]

Removes ebl's own stopped build containers (the disposable per-build containers
` + "`ebl build`" + ` creates, left behind if a build was interrupted or the client
disconnected before cleanup). With --all, also removes the shared gradle/npm
cache volumes and the runner/orchestrator/web images ` + "`ebl setup`/`ebl build`" + `
pulled - the next build/setup just re-pulls/re-creates whatever it needs, so
this is safe, just slower on the next run.

Options:
      --all                      Also remove cache volumes and pulled images,
                                  not just stopped containers
      --gradle-cache-volume <n>  Gradle cache volume name (default:
                                  expo-builder-local_gradle-cache - must match
                                  what ` + "`ebl build`" + ` used, if you customized it)
      --npm-cache-volume <n>     npm cache volume name (default:
                                  expo-builder-local_npm-cache)
      --docker-socket <path>     Docker socket path (default: /var/run/docker.sock;
                                  ignored on Windows)
  -h, --help                     Show this help
`

// PrintCleanUsage prints `ebl clean`'s help text.
func PrintCleanUsage() { fmt.Print(cleanUsage) }

// RunClean implements `ebl clean [options]`. args are the arguments after
// the "clean" subcommand token. Returns the process exit code.
func RunClean(args []string) int {
	all := false
	gradleCacheVolume := "expo-builder-local_gradle-cache"
	npmCacheVolume := "expo-builder-local_npm-cache"
	dockerSocket := "/var/run/docker.sock"

	needValue := func(i *int, flagName string) (string, bool) {
		if *i+1 >= len(args) {
			fmt.Fprintln(os.Stderr, color.Red("Missing value for "+flagName))
			return "", false
		}
		*i++
		return args[*i], true
	}

	for i := 0; i < len(args); i++ {
		arg := args[i]
		switch arg {
		case "-h", "--help":
			PrintCleanUsage()
			return 0
		case "--all":
			all = true
		case "--gradle-cache-volume":
			v, ok := needValue(&i, "--gradle-cache-volume")
			if !ok {
				return 2
			}
			gradleCacheVolume = v
		case "--npm-cache-volume":
			v, ok := needValue(&i, "--npm-cache-volume")
			if !ok {
				return 2
			}
			npmCacheVolume = v
		case "--docker-socket":
			v, ok := needValue(&i, "--docker-socket")
			if !ok {
				return 2
			}
			dockerSocket = v
		default:
			fmt.Fprintln(os.Stderr, color.Red("Unknown option: "+arg))
			return 2
		}
	}

	ctx := context.Background()
	docker := dockerapi.New(dockerSocket)
	if !docker.Ping(ctx) {
		fmt.Fprintln(os.Stderr, color.Red("Docker isn't reachable - is it running?"))
		return 1
	}

	fmt.Println(color.Bold("Removing stopped build containers..."))
	containers, err := docker.ListBuildContainers(ctx)
	if err != nil {
		fmt.Fprintln(os.Stderr, color.Red(err.Error()))
		return 1
	}

	anyRunning := false
	removedContainers := 0
	for _, c := range containers {
		if c.State == "running" {
			anyRunning = true
			continue
		}
		if err := docker.RemoveContainer(ctx, c.ID); err != nil {
			shortID := c.ID
			if len(shortID) > 12 {
				shortID = shortID[:12]
			}
			fmt.Println(color.Yellow("  Could not remove " + shortID + ": " + err.Error()))
		} else {
			removedContainers++
		}
	}
	fmt.Printf("  Removed %d container(s).\n", removedContainers)

	if !all {
		fmt.Println()
		fmt.Println(color.Green("Done.") + " Run with " + color.Cyan("--all") + " to also clear cache volumes and pulled images.")
		return 0
	}

	// Refuse rather than force a live build's cache volume out from under it -
	// Docker's own DELETE /volumes already fails on an in-use volume, but
	// checking here gives a clearer message than surfacing that raw error to
	// the user.
	if anyRunning {
		fmt.Println()
		fmt.Fprintln(os.Stderr, color.Red("A build is currently running - refusing to remove cache volumes/images out from under it. "+
			"Wait for it to finish (or stop it) and re-run `ebl clean --all`."))
		return 1
	}

	fmt.Println()
	fmt.Println(color.Bold("Removing cache volumes..."))
	for _, vol := range []string{gradleCacheVolume, npmCacheVolume} {
		if err := docker.RemoveVolume(ctx, vol); err != nil {
			fmt.Println(color.Yellow("  " + err.Error()))
		} else {
			fmt.Printf("  Removed volume %q.\n", vol)
		}
	}

	cfg, found, err := config.Load()
	if err != nil || !found {
		cfg = config.New()
	}
	fmt.Println()
	fmt.Println(color.Bold("Removing pulled images..."))
	for _, tag := range []string{cfg.RunnerImage(), cfg.OrchestratorImage(), cfg.WebImage()} {
		if err := docker.RemoveImage(ctx, tag); err != nil {
			fmt.Println(color.Yellow("  " + err.Error()))
		} else {
			fmt.Printf("  Removed image %q.\n", tag)
		}
	}

	fmt.Println()
	fmt.Println(color.Green(color.Bold("Done.")) + " The next `ebl build`/`ebl setup` will re-pull/re-create whatever it needs.")
	return 0
}

File: cli-cpp-legacy/src/commands/clean.cpp
#include "clean.hpp"

#include <iostream>
#include <string>

#include "../color.hpp"
#include "../config_store.hpp"
#include "../docker_client.hpp"

namespace ebl::commands {

namespace {

void printUsage() {
  std::cout << R"(ebl clean [options]

Removes ebl's own stopped build containers (the disposable per-build containers
`ebl build` creates, left behind if a build was interrupted or the client
disconnected before cleanup). With --all, also removes the shared gradle/npm
cache volumes and the runner/orchestrator/web images `ebl setup`/`ebl build`
pulled — the next build/setup just re-pulls/re-creates whatever it needs, so
this is safe, just slower on the next run.

Options:
      --all                      Also remove cache volumes and pulled images,
                                  not just stopped containers
      --gradle-cache-volume <n>  Gradle cache volume name (default:
                                  expo-builder-local_gradle-cache — must match
                                  what `ebl build` used, if you customized it)
      --npm-cache-volume <n>     npm cache volume name (default:
                                  expo-builder-local_npm-cache)
      --docker-socket <path>     Docker socket path (default: /var/run/docker.sock;
                                  ignored on Windows)
  -h, --help                     Show this help
)";
}

}  // namespace

void printCleanUsage() { printUsage(); }

int runClean(int argc, char** argv) {
  bool all = false;
  std::string gradleCacheVolume = "expo-builder-local_gradle-cache";
  std::string npmCacheVolume = "expo-builder-local_npm-cache";
  std::string dockerSocket = "/var/run/docker.sock";

  auto needValue = [&](int& i, const char* flagName) -> std::string {
    if (i + 1 >= argc) throw std::runtime_error(std::string("Missing value for ") + flagName);
    return argv[++i];
  };

  try {
    for (int i = 0; i < argc; i++) {
      std::string arg = argv[i];
      if (arg == "-h" || arg == "--help") {
        printUsage();
        return 0;
      }
      if (arg == "--all") { all = true; continue; }
      if (arg == "--gradle-cache-volume") { gradleCacheVolume = needValue(i, "--gradle-cache-volume"); continue; }
      if (arg == "--npm-cache-volume") { npmCacheVolume = needValue(i, "--npm-cache-volume"); continue; }
      if (arg == "--docker-socket") { dockerSocket = needValue(i, "--docker-socket"); continue; }
      std::cerr << ebl::color::red("Unknown option: " + arg) << "\n";
      return 2;
    }
  } catch (const std::exception& e) {
    std::cerr << ebl::color::red(e.what()) << "\n";
    return 2;
  }

  ebl::DockerClient docker(dockerSocket);
  if (!docker.ping()) {
    std::cerr << ebl::color::red("Docker isn't reachable — is it running?") << "\n";
    return 1;
  }

  std::cout << ebl::color::bold("Removing stopped build containers...") << "\n";
  std::vector<ebl::DockerClient::BuildContainerInfo> containers;
  try {
    containers = docker.listBuildContainers();
  } catch (const std::exception& e) {
    std::cerr << ebl::color::red(e.what()) << "\n";
    return 1;
  }

  bool anyRunning = false;
  int removedContainers = 0;
  for (const auto& c : containers) {
    if (c.state == "running") {
      anyRunning = true;
      continue;
    }
    try {
      docker.removeContainer(c.id);
      removedContainers++;
    } catch (const std::exception& e) {
      std::cout << ebl::color::yellow("  Could not remove " + c.id.substr(0, 12) + ": " + e.what()) << "\n";
    }
  }
  std::cout << "  Removed " << removedContainers << " container(s).\n";

  if (!all) {
    std::cout << "\n"
              << ebl::color::green("Done.") << " Run with " << ebl::color::cyan("--all")
              << " to also clear cache volumes and pulled images.\n";
    return 0;
  }

  // Refuse rather than force a live build's cache volume out from under it —
  // Docker's own DELETE /volumes already fails on an in-use volume, but checking
  // here gives a clearer message than surfacing that raw error to the user.
  if (anyRunning) {
    std::cerr << "\n"
              << ebl::color::red(
                     "A build is currently running — refusing to remove cache volumes/images out from under it. "
                     "Wait for it to finish (or stop it) and re-run `ebl clean --all`.")
              << "\n";
    return 1;
  }

  std::cout << "\n" << ebl::color::bold("Removing cache volumes...") << "\n";
  for (const auto& vol : {gradleCacheVolume, npmCacheVolume}) {
    try {
      docker.removeVolume(vol);
      std::cout << "  Removed volume \"" << vol << "\".\n";
    } catch (const std::exception& e) {
      std::cout << ebl::color::yellow(std::string("  ") + e.what()) << "\n";
    }
  }

  auto cfg = ebl::loadConfig().value_or(ebl::EblConfig{});
  std::cout << "\n" << ebl::color::bold("Removing pulled images...") << "\n";
  for (const auto& tag : {cfg.runnerImage(), cfg.orchestratorImage(), cfg.webImage()}) {
    try {
      docker.removeImage(tag);
      std::cout << "  Removed image \"" << tag << "\".\n";
    } catch (const std::exception& e) {
      std::cout << ebl::color::yellow(std::string("  ") + e.what()) << "\n";
    }
  }

  std::cout << "\n"
            << ebl::color::green(ebl::color::bold("Done."))
            << " The next `ebl build`/`ebl setup` will re-pull/re-create whatever it needs.\n";
  return 0;
}

}  // namespace ebl::commands
